packages/grpc-native-core/src/client_interceptors.js

```diff
@@ -0,0 +1,1343 @@
+    context;
+  if (this.delegate.onReceiveMessage) {
+    this.delegate.onReceiveMessage(message, next_method, context);
+  } else {
+    next_method(message);
+  }
+};
+
+/**
+ * Inbound status receiver.
+ * @param {grpc~StatusObject} status
+ */
+InterceptingListener.prototype.onReceiveStatus = function(status) {
+  this._callNext('onReceiveStatus', [status]);
+};
+
+/**
+ * A dead-end listener used to terminate a call chain. Used when an interceptor
+ * creates a branch chain, when the branch returns the listener chain will
+ * terminate here.
+ * @constructor
+ */
+function EndListener() {}
+EndListener.prototype.onReceiveMetadata = function(){};
+EndListener.prototype.onReceiveMessage = function(){};
+EndListener.prototype.onReceiveStatus = function(){};
+EndListener.prototype.recvMessageWithContext = function(){};
+
+/**
+ * Produces a callback triggered by streaming response messages.
+ * @private
+ * @param {EventEmitter} emitter
+ * @param {grpc.internal~Call} call
+ * @param {grpc~Listener} listener
+ * @param {grpc~deserialize} deserialize
+ * @return {Function}
+ */
+function _getStreamReadCallback(emitter, call, listener, deserialize) {
+  return function (err, response) {
+    if (err) {
+      // Something has gone wrong. Stop reading and wait for status
+      emitter.finished = true;
+      emitter._readsDone();
+      return;
+    }
+    var data = response.read;
+    var deserialized;
+    try {
+      deserialized = deserialize(data);
+    } catch (e) {
+      emitter._readsDone({
+        code: constants.status.INTERNAL,
+        details: 'Failed to parse server response'
+      });
+      return;
+    }
+    if (data === null) {
+      emitter._readsDone();
+      return;
+    }
+    var context = {
+      call: call,
+      listener: listener
+    };
+    listener.recvMessageWithContext(context, deserialized);
+  };
+}
+
+/**
+ * Produces an interceptor which will start gRPC batches for unary calls.
+ * @private
+ * @param {grpc~MethodDefinition} method_definition
+ * @param {grpc.Channel} channel
+ * @param {EventEmitter} emitter
+ * @param {function} callback
+ * @return {Interceptor}
+ */
+function _getUnaryInterceptor(method_definition, channel, emitter, callback) {
+  var serialize = method_definition.requestSerialize;
+  var deserialize = method_definition.responseDeserialize;
+  return function (options) {
+    var call = common.getCall(channel, method_definition.path, options);
+    var first_listener;
+    return new InterceptingCall(null, {
+      start: function (metadata, listener) {
+        var batch = {
+          [grpc.opType.SEND_INITIAL_METADATA]:
+            metadata._getCoreRepresentation(),
+        };
+        first_listener = listener;
+        call.startBatch(batch, function () {});
+      },
+      sendMessage: function(message) {
+        var batch = {
+          [grpc.opType.SEND_MESSAGE]: serialize(message),
+        };
+        call.startBatch(batch, function () {});
```

Comment:
How should I reconcile multiple batches (each with their own errors potentially) vs a single callback provided by the consumer? I've skipped the callback in early batches in favor of using it when receiving a status as seen in the [client streaming RPC](https://github.com/grpc/grpc-node/blob/master/packages/grpc-native-core/src/client.js#L645) but will this obscure useful errors from the consumer?

Reply:
I believe that what you are doing is correct. I think that every instance of multiple operations in a batch where the single callback actually matters includes a `status` operation. And I'm pretty sure that the status operation is always guaranteed to complete eventually. And I'm also pretty sure that if there is an error with another operation, it is always because of some circumstance that would cause a non-OK status. So in the end, you will want to surface the status.

One thing you need to be careful of here to avoid spurious errors is that you have to start the `SEND_MESSAGE` batch before you start the `SEND_CLOSE_FROM_CLIENT` batch, because the latter indicates that no more messages will be sent and the write will fail if you do them in the wrong order. I haven't seen yet if you are ensuring that ordering elsewhere.

Reply:
The non-streaming operations are ordered in client.js, here's the unary case: https://github.com/grpc/grpc-node/pull/59/files#diff-86af358fb1304514580c24212de4f182R486

This doesn't guarantee against a badly-writtten interceptor forcing an incorrect order but I assume we don't need to guard against that.

Reply:
Do you consider an interceptor that does any asynchronous operations to be badly written? Because I'm pretty sure that an asychronous write interceptor is sufficient to cause those operations to be performed out of order if there is no guard.

Reply:
This actually brings up another question for me: if a `sendMessage` interceptor on a streaming request takes an asnychronous action before passing the message to the next layer, is there any protection against messages getting passed down out of order?

Reply:
OK, the `sendMessageWithContext` actually answers that question, and I think it also provides a solution to the other problem. I am still worried about `start`, and not just because of unary calls. For any call, sending a message before sending the metadata will not do the right thing, and may crash.

Reply:
I don't think it's desirable to prevent an interceptor from changing the order of streaming messages. Interceptors can modify messages, swap a message out for a new message, or short circuit the operation entirely to skip a message. Building an ordered list of messages as they are passed in by the client's consumer and then enforcing that order when starting batches would make certain use cases impossible, like circuit breaking on streaming messages.

re: enforcing the order of operations in non-streaming requests, I was expecting interceptor authors to do that work if they want to make a continuation asynchronous. Adding a mechanism to the framework to delay an operation until a required operation is ready could reduce the burden on interceptor authors with async continuations. But it could also lead to difficult-to-debug scenarios with silent failures:

1) An interceptor author writes a bad interceptor which unexpectedly fails to call the SEND_MESSAGE continuation n% of the time.
2) An operation-order-enforcement mechanism in the framework allows the SEND_METADATA operation to be sent but doesn't send CLOSE n% of the time because it is waiting on a SEND_MESSAGE that never comes.
3) n% of all calls pile up without error until they time out or memory runs out.

The same bad interceptor would cause immediate failures with useful errors if the framework is not waiting on an order-enforcement mechanism.

Reply:
The use of `sendMessageWithContext` already enforces the order of messages by preventing the top-level stream from sending down each message until the previous message has been fully processed.

And from my point of view, it feels very surprising to users that this interceptor API is clearly written to support asynchronous operations, but if they try to actually use an asynchronous operation, it will screw up some requests unless they do significant extra work to explicitly track their own pending operations. On top of that, in the current library it is not possible for the user to call the API in a way that results in out-of-order calls to the C core. I think that is a valuable guarantee and I would be very hesitant to change it.